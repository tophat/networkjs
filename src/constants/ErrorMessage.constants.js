import NetworkStatuses from './NetworkStatus.constants'

export const ErrorMessage = {
    INVALID_EVENT: `Event must be one of ${NetworkStatuses}`,
}

export const ErrorMessages = Object.values(ErrorMessage)
