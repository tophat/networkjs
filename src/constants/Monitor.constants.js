export const Monitor = {
    NETWORK: 'network',
    SERVICE: 'service',
    STABILITY: 'stability',
}

export const Monitors = Object.values(Monitor)
