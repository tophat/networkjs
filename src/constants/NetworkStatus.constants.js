export const NetworkStatus = {
    ONLINE: 'online',
    OFFLINE: 'offline',
    STABLE: 'stable',
    UNSTABLE: 'unstable',
    DEGRADED: 'degraded',
    RESOLVED: 'resolved',
}

export const NetworkStatuses = Object.values(NetworkStatus)
