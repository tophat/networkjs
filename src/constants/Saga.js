export const Saga = {
    NETWORK: 'network',
    SERVICE: 'service',
    STABILITY: 'stability',
}

export const Sagas = Object.values(Saga)
