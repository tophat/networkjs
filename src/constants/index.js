export { NetworkStatus, NetworkStatuses } from './NetworkStatus.constants'
export { Monitor, Monitors } from './Monitor.constants'
export { ErrorMessage, ErrorMessages } from './ErrorMessage.constants'
