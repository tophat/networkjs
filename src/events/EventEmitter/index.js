import Event from '../Event'

export default class EventEmitter {
    constructor() {
        this.events = {}
    }

    registerEvent(eventName) {
        const event = new Event(eventName)
        this.events[eventName] = event
    }

    dispatchEvent(eventName, ...eventArgs) {
        this.events[eventName].callbacks.forEach(callback => {
            callback(...eventArgs)
        })
    }

    addEventListener(eventName, callback) {
        this.events[eventName].registerCallback(callback)
    }
}
