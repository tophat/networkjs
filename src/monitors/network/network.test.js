import EventEmitter from '../../events/EventEmitter'
import NetworkMonitor from '.'
import { NetworkStatus } from '../../constants'

jest.mock('../../events/EventEmitter')

describe('Network Monitor', () => {
    describe('NetworkMonitor', () => {
        let monitor
        const emitter = new EventEmitter()

        beforeEach(() => {
            monitor = new NetworkMonitor(emitter)
        })

        afterEach(() => {
            jest.clearAllMocks()
        })

        describe('constructor', () => {
            it('initializes with the correct props', () => {
                expect(monitor).toMatchSnapshot()
            })
        })

        describe('initialize', () => {
            it('adds two event listeners to the window', () => {
                window.addEventListener = jest.fn()
                monitor.initialize()

                expect(window.addEventListener).toHaveBeenCalledTimes(2)
                expect(window.addEventListener).toHaveBeenNthCalledWith(
                    1,
                    'online',
                    monitor.emitOnlineEvent,
                )
                expect(window.addEventListener).toHaveBeenNthCalledWith(
                    2,
                    'offline',
                    monitor.emitOfflineEvent,
                )
            })
        })

        describe('pause', () => {
            it('removes two event listeners to the window', () => {
                window.removeEventListener = jest.fn()
                monitor.pause()

                expect(window.removeEventListener).toHaveBeenCalledTimes(2)
            })
        })

        describe('resume', () => {
            it('runs initalize which adds two event listeners', () => {
                window.addEventListener = jest.fn()
                const initializeSpy = jest.spyOn(monitor, 'initialize')
                monitor.resume()

                expect(window.addEventListener).toHaveBeenCalledTimes(2)
                expect(initializeSpy).toHaveBeenCalled()
            })
        })

        describe('event emitter', () => {
            it('dispatches online event on window online', () => {
                monitor.emitOnlineEvent()

                expect(monitor.emitter.dispatchEvent).toHaveBeenCalledWith(
                    NetworkStatus.ONLINE,
                )
            })
            it('dispatches offline event on window offline', () => {
                monitor.emitOfflineEvent()

                expect(monitor.emitter.dispatchEvent).toHaveBeenCalledWith(
                    NetworkStatus.OFFLINE,
                )
            })
        })
    })
})
