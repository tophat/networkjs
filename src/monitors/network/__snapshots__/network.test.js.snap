// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network Monitor NetworkMonitor constructor initializes with the correct props 1`] = `
NetworkMonitor {
  "emitOfflineEvent": [Function],
  "emitOnlineEvent": [Function],
  "emitter": EventEmitter {
    "addEventListener": [MockFunction],
    "dispatchEvent": [MockFunction],
    "registerEvent": [MockFunction],
  },
}
`;
